File: backend/models/perusahaan.py
from sqlalchemy.exc import IntegrityError

from rpc.gen.data.perusahaan.structs.ttypes import TPerusahaanForm, TPerusahaan
from rpc.gen.data.perusahaan.errors.ttypes import TPerusahaanError, TPerusahaanErrorCode
from rpc.gen.user.user.types.ttypes import TUserRole

import db
from db.entities.staging import DbPerusahaan
from db.errors import UniqueError
from .crud import CrudModel

class PerusahaanModel(CrudModel):
    def __init__(self):
        CrudModel.__init__(
            self, 
            DbPerusahaan, 
            TPerusahaanError, 
            TPerusahaanErrorCode,
            "PERUSAHAAN", 
            name_field="name", 
            actions={
                "get": TUserRole.ADMIN_BIASA,
                "create": TUserRole.ADMIN_UTAMA,
                "fetch": TUserRole.ADMIN_BIASA,
                "delete": TUserRole.SUPER_ADMIN,
            },
            setters={
                "name": TUserRole.ADMIN_UTAMA,
                "enabled": TUserRole.ADMIN_UTAMA,
            },
        )

File: frontend/src/lib/validators/data/perusahaan.js
import { 
	TPerusahaanErrorCode, T_PERUSAHAAN_ERROR_STR,
} from '@/rpc/gen/data.perusahaan.errors_types';
import errors from '@/rpc/gen/data.perusahaan.errors_types';
import types from '@/rpc/gen/user.user.types_types';
import { createRulesField as _createRulesField, createRulesFields } from "@/lib/validators/common";

export const createRulesField = (field) => _createRulesField(
	field,
	errors,
	types, 
	TPerusahaanErrorCode, 
	T_PERUSAHAAN_ERROR_STR, 
	"PERUSAHAAN"
);
export const NAME_RULES = createRulesField("NAME");

const RULES = createRulesFields(
	["NAME"],
	errors,
	types, 
	TPerusahaanErrorCode, 
	T_PERUSAHAAN_ERROR_STR, 
	"PERUSAHAAN"
);
export default RULES;
